src/util/mpsc.rs

```diff
@@ -190,76 +208,87 @@ pub fn bounded<T>(cap: usize) -> (Sender<T>, Receiver<T>) {
     )
 }
 
-impl<T> Stream for Receiver<T> {
-    type Error = ();
-    type Item = T;
-
-    fn poll(&mut self) -> Poll<Option<T>, ()> {
-        match self.try_recv() {
-            Ok(m) => Ok(Async::Ready(Some(m))),
-            Err(TryRecvError::Empty) => {
-                let t = self.state.recv_task.swap(task::current(), Ordering::AcqRel);
-                if t.is_some() {
-                    return Ok(Async::NotReady);
-                }
-                // Note: If there is a message or all the senders are dropped after
-                // polling but before task is set, `t` should be None.
-                self.poll()
-            }
-            Err(TryRecvError::Disconnected) => Ok(Async::Ready(None)),
-        }
-    }
-}
-
 const CHECK_INTERVAL: usize = 8;
 
 /// A sender of channel that limits the maximun pending messages count loosely.
 pub struct LooseBoundedSender<T> {
     sender: Sender<T>,
-    tried_cnt: usize,
+    tried_cnt: Cell<usize>,
     limit: usize,
 }
 
 impl<T> LooseBoundedSender<T> {
+    /// Returns the number of messages in the channel.
+    #[inline]
+    pub fn len(&self) -> usize {
+        self.sender.len()
+    }
+
+    /// Returns true if the channel is empty.
+    ///
+    /// Note: Zero-capacity channels are always empty.
+    #[inline]
+    pub fn is_empty(&self) -> bool {
+        self.sender.is_empty()
+    }
+
     /// Send a message regardless its capacity limit.
     #[inline]
-    pub fn force_send(&mut self, t: T) -> Result<(), SendError<T>> {
-        self.tried_cnt += 1;
+    pub fn force_send(&self, t: T) -> Result<(), SendError<T>> {
```

Comment:
Why not `&mut self`? I think it's the common style. And `Cell` make the `Sender` not `Sync` any more.

Reply:
It's intended. Actually, `&self` is very common for sending a message. And not using a mutable reference is useful for integrating it into existing transport system.

Reply:
Ok. So do you think `impl !Sync for LooseBoundedSender` explicitly is a good idea? Just make the doc more friendly.

Reply:
That would be unnecessary. `!Sync` is not intended. Internal mutability is intended. It's a quite common in rust world that sender is not Sync as sender in std library is not Sync.